### ISSUE TITLE
fixedDisplacementZeroShear extremely slow to converge

### ISSUE THREAD

**Issue by denMaier** [solids4foam/solids4foam#132]
Hey,

for my geotechnical calculations it is almost always necessary to initialize the stress state linear-elastically before using a plastic material law. 
For that I let the soil settle under its own weight. See the minimal example attached and in this picture:
![Bildschirmfoto vom 2024-06-25 17-23-26](https://github.com/solids4foam/solids4foam/assets/73213965/2d520a36-64f6-4fb3-9897-933809c9a0bb)
[simpleCubeWeight.zip](https://github.com/user-attachments/files/15973758/simpleCubeWeight.zip)

However, i noticed that the main reason holding the convergence back is the 'fixedDisplacementZeroShear' BC. 
With linear elasticity most of the domain should converge within one outer iteration. We only need multiple iterations because the boundary conditions are differentiated and updated after the linear solver has converged.
However, the top boundary (zero traction) converged quite fast. For the attached case, its obviously not too bad, since there are not a lot of cells, however for larger cases it gets quite extreme and the convergence somehow becomes slower with mesh refinement. Also the attached case is pretty small in dimensions, so the residual shear stresses are small. In the larger cases I get shear stresses in the 100s kPa.. 

I also noticed that the problem is worse with tetMesh than with hexMesh (although, adding a hex boundary layer to the tetMesh made it somehow worse).

Any help would be highly appreciated!

**Comment by iBatistic** [comment]
Hi @denMaier,
> With linear elasticity most of the domain should converge within one outer iteration. We only need multiple iterations because the boundary conditions are differentiated and updated after the linear solver has converged.

This is true only for the block-coupled solver, where each term is implicitly discretized. The solver you are using (`linearGeometryTotalDisplacement`) is semi-implicit, meaning that the Laplacian trace and Laplacian transpose are discretised explicitly.

I checked your `fvSchemes` dictionary. Did you try using the standard schemes we typically use in solid cases? I tested standard ones with a 20x20 mesh, and it converged in 23 outer iterations. With a 500x500 mesh, it converged in 13 outer iterations.

I have never tried the `iterativeGauss` scheme. What has your experience been with it? Are you getting solutions close to those obtained with the `leastSquares` scheme?

**Comment by denMaier** [comment]
thanks for the reply @iBatistic,
I had very good experiences with the iterative Gauss in cases with skewed and non-orthogonal meshes to remedy the oscillations arising. However, I will try my large case with the standard schemes and update this thread. 

Maybe to use this as an opportunity to fill some holes in my understanding: 
> This is true only for the block-coupled solver, where each term is implicitly discretized. The solver you are using (linearGeometryTotalDisplacement) is semi-implicit, meaning that the Laplacian trace and Laplacian transpose are discretised explicitly.
Is this because we use the scalar kImp (which is indecently the the p-wave or oedometric stiffness in geo-dynamics/mechanics) instead of the complete elastoplastic stiffness tensor in the laplacian(kImp,D) terms?

**Comment by denMaier** [comment]
Ok so, changing back to the standard schemes did speadup the calculation, which is almost expected since iterative Gauss is quite expensive. However I still need the same 668 outer iterations to get the boundaries to a state, where they at least don't introduce noticable shear stresses compared to the main problem (an embankment).  
Would you suggest me to change to a block coupled solver? Also do you have any experience with the block-coupled linear solvers in the ESI versions? 
In general, maybe it would be a good idea to add a discussion to this repo about which solidModels are available and when should they be chosen over others? I feel like there is a lot happening in solids4Foam but I can't quite catch up.

**Comment by iBatistic** [comment]
> Is this because we use the scalar kImp (which is the the p-wave or oedometric stiffness in geo-dynamics/mechanics) instead of the complete elastoplastic stiffness tensor in the laplacian(kImp,D) terms?

You can say that. Philip explained it very nicely in the article titled "An Open-Source Finite Volume Toolbox for Solid Mechanics and Fluid-Solid Interaction Simulations." When using the segregated approach, where the coefficients in the stiffness matrix are scalars, we can only implicitly discretize the first term (the Laplacian, highlighted in green). The second and third terms, marked in red, are Laplacian trace and Laplacian transpose. Their implicit discretization results in coefficients in the matrix that are tensors, requiring a block solver. I assume you are familiar with these concepts, but I am including them here for the benefit of others that may read this discussion.

![Screenshot from 2024-06-26 14-07-43](https://github.com/solids4foam/solids4foam/assets/19718956/2c8d30e5-cd53-43bd-ac37-c15781df7df2)

> However I still need the same 668 outer iterations to get the boundaries to a state, where they at least don't introduce noticable shear stresses compared to the main problem (an embankment).

Hmmmm strange, can you check the attached schemes I used (I run the case with fe41). I don't have any sigma_xy stress at boundaries.
[schemes.zip](https://github.com/user-attachments/files/15988128/schemes.zip)


> Would you suggest me to change to a block coupled solver? Also do you have any experience with the block-coupled linear solvers in the ESI versions?

Block solver for linear elasticity is currently only implemented with the foam-extend version. Since the majority of the code is designed for the segregated approach, I suggest you stick with that solvers if possible.

**Comment by denMaier** [comment]
Thank you a lot for the refresher, @iBatistic . I also went back to @philipcardiff's article. I knew this once, but it's been a while since i read the article the last time..

I found that, indeed, using "extendedLeastSquares 0" for the gradients now does enhance convergence for the boundaries considerably. I don't know why my last tests didn't seem to show any improvements. If I can find the time I will do a little study on extendedLeastSquares and iterative gauss on non ideal meshes to see where the strengths of both lie.
